mm2src/common/custom_futures/repeatable.rs

```diff
@@ -0,0 +1,638 @@
+
+pub trait FactoryTrait<F>: Fn() -> F {}
+
+impl<Factory, F> FactoryTrait<F> for Factory where Factory: Fn() -> F {}
+
+pub trait RepeatableTrait<T, E>: Future<Output = Action<T, E>> + Unpin {}
+
+impl<F, T, E> RepeatableTrait<T, E> for F where F: Future<Output = Action<T, E>> + Unpin {}
+
+pub(crate) trait InspectErrorTrait<E>: 'static + Fn(&E) + Send {}
+
+impl<F: 'static + Fn(&E) + Send, E> InspectErrorTrait<E> for F {}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum RepeatError<E> {
+    TimeoutExpired {
+        until_ms: u64,
+        /// An error occurred during the last attempt.
+        error: E,
+    },
+    AttemptsExceed {
+        attempts: usize,
+        /// An error occurred during the last attempt.
+        error: E,
+    },
+}
+
+impl<E> RepeatError<E> {
+    pub fn error(&self) -> &E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    pub fn into_error(self) -> E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    fn timeout(until_ms: u64, error: E) -> Self { RepeatError::TimeoutExpired { until_ms, error } }
+
+    fn attempts(attempts: usize, error: E) -> Self { RepeatError::AttemptsExceed { attempts, error } }
+}
+
+impl<E: fmt::Display> fmt::Display for RepeatError<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RepeatError::TimeoutExpired { until_ms, error } => {
+                write!(
+                    f,
+                    "Waited too long until {until_ms}ms for the future to succeed. Error: {error}",
+                )
+            },
+            RepeatError::AttemptsExceed { attempts, error } => {
+                write!(f, "Error {error} on retrying the future after {attempts} attempts")
+            },
+        }
+    }
+}
+
+/// The future is ether ready (with a `T` result), or not ready (failed with an intermediate `E` error).
+#[derive(Debug)]
+pub enum Action<T, E> {
+    Ready(T),
+    Retry(E),
+}
+
+pub trait RetryOnError<T, E> {
+    fn retry_on_err(self) -> Action<T, E>;
+}
+
+impl<T, E> RetryOnError<T, E> for Result<T, E> {
+    /// Converts `Result<T, E>` into `Action<T, E>`:
+    /// * `Ok(T)` => `Action::Ready(T)`.
+    /// * `Err(E)` => `Action::Retry(E)`.
+    #[inline]
+    fn retry_on_err(self) -> Action<T, E> {
+        match self {
+            Ok(ready) => Action::Ready(ready),
+            Err(e) => Action::Retry(e),
+        }
+    }
+}
+
+/// The result of `repeatable` or `retry_on_err` macros - the first step at the future configuration.
+pub struct Repeatable<Factory, F, T, E> {
+    factory: Factory,
+    /// Currently executable future, i.e. an active attempt.
+    exec_fut: F,
+    /// A timeout future if we're currently waiting for a timeout.
+    timeout_fut: Option<Timer>,
+    until: Option<RepeatUntil>,
```

Comment:
With the latest changes, this can be safely made non-`Option` with a default value. Please consider this diff (all repeatable tests pass with it):
```diff
--- a/mm2src/common/custom_futures/repeatable.rs
+++ b/mm2src/common/custom_futures/repeatable.rs
@@ -186,7 +186,7 @@ pub struct Repeatable<Factory, F, T, E> {
     exec_fut: F,
     /// A timeout future if we're currently waiting for a timeout.
     timeout_fut: Option<Timer>,
-    until: Option<RepeatUntil>,
+    until: RepeatUntil,
     repeat_every: Duration,
     inspect_err: Option<Box<dyn InspectErrorTrait<E>>>,
     _phantom: PhantomData<(F, T, E)>,
@@ -205,7 +205,10 @@ where
             factory,
             exec_fut,
             timeout_fut: None,
-            until: None,
+            until: RepeatUntil::AttemptsExceed(AttemptsState {
+                current_attempt: 0,
+                total_attempts: 1,
+            }),
             repeat_every: DEFAULT_REPEAT_EVERY,
             inspect_err: None,
             _phantom: PhantomData::default(),
@@ -246,7 +249,7 @@ where
     pub fn attempts(mut self, total_attempts: usize) -> Self {
         assert!(total_attempts > 0, "'total_attempts' cannot be 0");
 
-        self.until = Some(RepeatUntil::AttemptsExceed(AttemptsState::new(total_attempts)));
+        self.until = RepeatUntil::AttemptsExceed(AttemptsState::new(total_attempts));
         self
     }
 
@@ -258,7 +261,7 @@ where
             warn!("Deadline has already passed: now={now:?} until={until_ms:?}")
         }
 
-        self.until = Some(RepeatUntil::TimeoutMsExpired(until_ms));
+        self.until = RepeatUntil::TimeoutMsExpired(until_ms);
         self
     }
 
@@ -306,22 +309,18 @@ where
                     }
 
                     match self.until {
-                        Some(RepeatUntil::TimeoutMsExpired(until_ms)) => {
+                        RepeatUntil::TimeoutMsExpired(until_ms) => {
                             if !self.check_can_retry_after_timeout(until_ms) {
                                 return Poll::Ready(Err(RepeatError::timeout(until_ms, error)));
                             }
                         },
-                        Some(RepeatUntil::AttemptsExceed(ref mut attempts)) => {
+                        RepeatUntil::AttemptsExceed(ref mut attempts) => {
                             // Check if we have one more attempt to retry to execute the future.
                             attempts.current_attempt += 1;
                             if attempts.current_attempt >= attempts.total_attempts {
                                 return Poll::Ready(Err(RepeatError::attempts(attempts.current_attempt, error)));
                             }
                         },
-                        // We should try to execute the future only once
-                        // if neither `Self::attempts` nor `Self::with_timeout` are specified.
-                        // https://github.com/KomodoPlatform/atomicDEX-API/pull/1564#discussion_r1050778208
-                        None => return Poll::Ready(Err(RepeatError::attempts(1, error))),
                     }
```

Reply:
I should have done this after this comment https://github.com/KomodoPlatform/atomicDEX-API/pull/1564#discussion_r1057181772
Thanks for the pointing! Hope everything is done now 🙂